### PR TITLE
[CSV Reader] Spinlock over GetLine Error + New Strategy for dialect candidates

### LINKED ISSUE #10749: Read csv hangs on imdb dataset
### What happens?

I try to load file with

```
select * from read_csv('title.principals.tsv.gz') limit 10;
```

and it hangs

file from here https://developer.imdb.com/non-commercial-datasets/

used to work in 0.9.2

### To Reproduce

Just download files from https://datasets.imdbws.com/title.principals.tsv.gz

and try to load the TSVs in Duck DB

### OS:

Windows 11

### DuckDB Version:

0.10.0

### DuckDB Client:

CLI

### Full Name:

Vasileios Anagnostopoulos

### Affiliation:

Agile Actors

### Have you tried this on the latest [nightly build](https://duckdb.org/docs/installation/?version=main)?

I have tested with a release build (and could not test with a nightly build)

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] Yes, I have

### PR DESCRIPTION
**Pull request by pdet**
Fix: #10749

This issue had 2 different problems.

1. Spinlock hanging
The dialect detection would pick quotes, that would later cause an `unquoted_value` error. However, because this query uses a `LIMIT` operator, there is no guarantee that previous boundaries, before the one the error occurs, would be ever finished, since they can be terminated before fully executing the boundary. This would cause the CSV Error Handler to wait for boundaries that would never be finished. To solve that, I've added an extra unordered_set that keeps track of boundaries that are still being processed, hopefully, this will also avoid any other errors like that.

2. Quote Dialect Detection
I've changed the rules on how we pick quotes. The basic idea is that if we end up with our dialect detection with multiple quote candidates, we either give preference to a quote option that has been used during the scan of at least one value in the file, otherwise we give preference to the null terminator.

Unfortunately, I did not manage to add the file as a test, since the original file is quite big. I've attempted to generate a smaller file, but the smallest I got was ~18Mb compressed, which seems like an overkill, but I'm open to suggestions.

### CODE REVIEW

**Comment by Mytherin** [comment]
> Unfortunately, I did not manage to add the file as a test, since the original file is quite big. I've attempted to generate a smaller file, but the smallest I got was ~18Mb compressed, which seems like an overkill, but I'm open to suggestions.

You could add the file here: https://github.com/duckdb/duckdb-data

File: src/execution/operator/csv_scanner/util/csv_error.cpp [hunk 1]
```diff
@@ -171,10 +182,15 @@ idx_t CSVErrorHandler::GetLine(LinesPerBoundary &error_info) {
...
 			}
 			lock_guard<mutex> parallel_lock(main_mutex);
+
 			if (lines_per_batch_map.find(boundary_idx) != lines_per_batch_map.end()) {
 				batch_done = true;
 				current_line += lines_per_batch_map[boundary_idx].lines_in_batch;
 			}
+			if (in_process.find(boundary_idx) == in_process.end()) {
+				// This boundary index is not being processed and also has not finished, we skip it.
+				batch_done = true;
```
[the remark this answers is not in the document]
Comment (pdet) [reply src/execution/operator/csv_scanner/util/csv_error.cpp:192]: I think that sounds like a good idea,
I probably need an ordered set alive with the boundary indexes that are still computing, and when they finish we just need to check if there is no boundary index smaller than the one that error-ed and then throw.

Thar would indeed remove the spinlock, which is also my fear that could cause issues we can't predict.

File: src/execution/operator/csv_scanner/util/csv_error.cpp [hunk 2]
```diff
@@ -13,26 +13,12 @@ LinesPerBoundary::LinesPerBoundary(idx_t boundary_idx_p, idx_t lines_in_batch_p)
...
 	std::ostringstream error;
 	if (PrintLineNumber(csv_error)) {
-		error << "CSV Error on Line: " << GetLine(error_info) << std::endl;
-	}
-	{
-		lock_guard<mutex> parallel_lock(main_mutex);
-		got_borked = true;
+		error << "CSV Error on Line: " << GetLine(csv_error.error_info) << std::endl;
 	}
+	{ lock_guard<mutex> parallel_lock(main_mutex); }
```
Comment (Mytherin) [inline src/execution/operator/csv_scanner/util/csv_error.cpp:21]: Is this intentional?

**Review by Mytherin — COMMENTED** [review]
Thanks for the updates! I played around with this a bit and it seems to deadlock in `CSVErrorHandler::Insert` here:

```sql
CALL dbgen(sf=1);
COPY (
  SELECT * REPLACE (
  CASE WHEN row_number() OVER ()=1234567 THEN 'thisisnotanumber'
  ELSE l_orderkey::VARCHAR
  END AS l_orderkey)
  FROM lineitem) TO 'lineitem_malformed.csv';
COPY lineitem FROM lineitem_malformed.csv;
```

May want to have a look at running the CSV error reporting with thread sanitizer, particularly for bigger files.

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**